src/core/server/csp/csp.ts

```diff
@@ -0,0 +1,86 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { TypeOf, schema } from '@kbn/config-schema';
+
+/**
+ * The CSP options used for Kibana.
+ * @public
+ */
+export interface CspOptions {
+  /**
+   * The CSP rules in a formatted directives string for use
+   * in a `Content-Security-Policy` header.
+   */
+  directives: string;
+
+  /**
+   * The CSP rules used for Kibana.
+   */
+  rules: string[];
+
+  /**
+   * Specify whether browsers that do not support CSP should be
+   * able to use Kibana. Use `true` to block and `false` to allow.
+   */
+  strict: boolean;
+
+  /**
+   * Specify whether users with legacy browsers should be warned
+   * about their lack of Kibana security compliance.
+   */
+  warnLegacyBrowsers: boolean;
+}
+
+export type CspConfigType = TypeOf<typeof config.schema>;
+
+const DEFAULT_RULES = [
+  `script-src 'unsafe-eval' 'self'`,
+  `worker-src blob: 'self'`,
+  `style-src 'unsafe-inline' 'self'`,
+];
+const defaults = {
+  directives: createCspDirectives(DEFAULT_RULES),
+  rules: DEFAULT_RULES,
+  strict: true,
+  warnLegacyBrowsers: true,
+};
+export const config = {
+  path: 'csp',
```

Comment:
Probably we should move this to `server.csp` in the future. Can you add a note to #50644 that we should rename this key using the config deprecations #52251 once it's available?

Reply:
Definitely! 👍

Reply:
I'm currently looking into adding support for additional security headers.

Whenever we do rename this, how do you both feel about something like `server.securityHeaders.csp`? Then the other headers can use the same prefix, and all related headers will be in one place.

Reply:
Should it be "security" specific? What about just `server.headers.csp`?

Reply:
I think both make sense, but with my AppSec hat on I lead towards being more specific. Reason being is so these settings stand out when people search for "security", since these settings would apply to core, not to the Security plugin. We won't be able to have the ideal defaults set for all of these (e.g., `X-Frame-Options`), so we'll want to ensure that users are aware these exist.

Also, I can't think of any other non-security-related HTTP headers that we'd want to have specific config settings for.

Reply:
> Also, I can't think of any other non-security-related HTTP headers that we'd want to have specific config settings for.

CORS? I'm fine if Security tams owns it